### PR TITLE
Convert Lombok `@Value` classes to Java records

### PR DESCRIPTION
**Pull request by kevin0x90**
## What's changed?
Adds a recipe for converting lombok value annotated classes to java records.

## What's your motivation?
I am working on a case study to evaluate the effort and benefit of automated refactorings.

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

**Comment by timtebeek** [comment]
This is looking very promising already! Thanks for getting this started. Could you run `./gradlew lF` to add the license headers? That would make the tests run here.

**Comment by kevin0x90** [comment]
> This is looking very promising already! Thanks for getting this started. Could you run `./gradlew lF` to add the license headers? That would make the tests run here.

Hi,

Sorry for the late reply and thank you for adding the copyright information.

I added more changes to approach the defined acceptability.
During the replacement of the getter methods I ran into some type problems. Maybe you have an idea how to prevent this?

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:86]: The default is also five minutes, so no need to override that here.
```suggestion
```

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:69]:
```suggestion
        return "Convert `@lombok.Value` class to Record";
```

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:74]:
```suggestion
        return "Convert Lombok `@Value` annotated classes to standard Java Records.";
```

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:96]: I'm doubtful if we need these overrides; were these added intentionally & necessary for these replacements?
```suggestion
```

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:100]: Perhaps good to limit recipe execution here early by checking for the presence of `lombok.Value` through something like 
```suggestion
        final TreeVisitor<?, ExecutionContext> check = Preconditions.and(
                new UsesJavaVersion<>(17),
                new UsesType("lombok.Value")
```
Reviewing from GitHub, so the above might still need an import for instance.

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:76]: Probably best to get this resolved before we merge, such that subsequent recipes don't trip over missing types; let us know if you'd need help to resolve the missing types!

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (timtebeek) [inline src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:40]: Since (nearly) all tests use this version, we can also set it in the `defaults` method above to simplify each test below a bit.

**Review by timtebeek — COMMENTED** [review]
Quick first round of feedback; really like that you took this on! Hope to get this closer to a merge soon, and start trying it out on projects.

Did you ever look towards using the Moderne CLI? That might allow you to run it locally across projects quickly, after a first round of building serialized LSTs. Not a requirement, but could give you feedback on how this works on your own projects without the repeated wait for the lossless semantic tree models to be built.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:341]: Should we additionally check if there are any fields annotated with `@With` for instance? Annotations on any field should disqualify the class as well I suppose; not sure if that's already correctly handled, and didn't spot a corresponding case in the (extensive!) tests yet.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:234]: Maybe good to pull this last one up, as it's cheaper to check than some of the others.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:96]: Yes currently these overrides are required since the recipe runs in two phases

- First run: adjust the value classes to records and keep track of the members 
- Second run: adjust the references to getters of the adjusted classes

I haven't found another way to solve the issue of adjusting the references in the same cycle because from what I have read in the documentation the recipe is run in parallel when executed outside of a test environment.

If there is a solution how to solve this with only one run I would be really interested in it.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:100]: I like the suggestion but since I need to run the recipe twice I am not sure if this would break things because from my understanding in the second run the lombok Value classes would be already adjusted and therefore the import would be missing.

I will look into it and adjust if possible.

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (kevin0x90) [reply src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:76]: I would be really happy to get support regarding this. At the moment i suspect in the test somehow even the types are in the same package somehow the environment does not know the adjusted types after the first run.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:341]: Sounds like a good idea. Maybe it would be good to filter for general field annotations related to lombok for this recipe.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (knutwannheden) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:96]: Enter `ScanningRecipe`: https://docs.openrewrite.org/concepts-explanations/recipes#scanning-recipes

Recipes extending this class will run in two phases: The first phase collects data across all sources and the second phase performs the required modifcations.

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (timtebeek) [reply src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:40]: In the `defaults` method you can call `.allSources(src -> src.markers(javaVersion(17)));` on the spec to set that marker on all sources.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:341]: I think just any field annotation would disqualify the conversion here; I'm not sure we'd correctly handle Jackson annotations either for instance, so I'd like to start out with the simplest case only: class with only fields and only a single annotation in total, such that we don't accidentally break any code with this recipe. We can always (safely) expand the scope of what we cover later; we can't easily recover trust when we break things. :)

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java]: Thanks for the hint, I refactored the implementation so that a ScanningRecipe is used.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java]: Addressed with the refactoring to ScanningRecipe

**Comment by kevin0x90** [comment]
> Quick first round of feedback; really like that you took this on! Hope to get this closer to a merge soon, and start trying it out on projects.
> 
> Did you ever look towards using the Moderne CLI? That might allow you to run it locally across projects quickly, after a first round of building serialized LSTs. Not a requirement, but could give you feedback on how this works on your own projects without the repeated wait for the lossless semantic tree models to be built.

I did not yet try out the Moderne CLI but will give it a try for my sample project.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:84]: Rather than checking the import we can instead check for `UsesType`. Otherwise we might fail to pick up qualified use through `@lombok.Value`.

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (timtebeek) [reply src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:78]: It seems you've worked through most of the review feedback, such that there's less chance for conflicts if we were to step in to help out here. Do you still need help? Typically I'd step through with the debugger and try things that way, but not sure yet as I haven't looked

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:93]: Ok, will adjust

**Comment by timtebeek** [comment]
Finally found some more time to review; haven't yet gotten to the missing type information, but did find some simplifications and a case not yet covered around static fields. Not quite sure what records allow there, but might be safest to just not convert those cases. I've also grouped the tests into a nested class, such that it's easier to spot what should and should not be supported.

Let me know if you'd like to fix that failing test yourself, or if you'd at any point want to hand this over. :)

**Comment by kevin0x90** [comment]
> Finally found some more time to review; haven't yet gotten to the missing type information, but did find some simplifications and a case not yet covered around static fields. Not quite sure what records allow there, but might be safest to just not convert those cases. I've also grouped the tests into a nested class, such that it's easier to spot what should and should not be supported.
> 
> Let me know if you'd like to fix that failing test yourself, or if you'd at any point want to hand this over. :)

Hi,

I will fix the failing test and keep on working on this branch whenever I find the time to do so.

File: src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java
Comment (timtebeek) [reply src/test/java/org/openrewrite/java/migrate/lombok/LombokValueToRecordTest.java:81]: So the type information is already present on the input, and we're reusing the method invocations from before. As such I don't think we need to hold up this PR over missing type information, as a next run of OpenRewrite using the same sources would actually work. We can leave the missing type information as a nice to have for a future enhancement, with limited use.

**Review by timtebeek — APPROVED** [review]
Looks good to me! Thanks again.

Any last minute feedback @knutwannheden or @joanvr ?

**Comment by knutwannheden** [comment]
I can take a look.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:252]: Rather than throwing an exception, it would be better if we just made sure, that the recipe doesn't do anything. Typically, a class should indeed by correctly type attributed. But when this recipe is run together with others, we can't always have too high expectations of what the other recipes do, so it could happen that another recipe gets that wrong. So that this gets detected during development, we could add an `assert` statement instead.

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java]: removed

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java]: addressed

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java]: addressed

File: src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java
Comment (kevin0x90) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:59]: from my perspective this would be fine to make the option non-required but personally I would prefer to keep a primitive boolean that is initialized from the default constructor to false so basically like:
```
@Option(displayName = "Add a `toString()` implementation matching Lombok",
            description = "When set the `toString` format from Lombok is used in the migrated record.",
            required = false)
boolean useExactToString;

public LombokValueToRecord() {
        this(false);
}

public LombokValueToRecord(boolean useExactToString) {
        this.useExactToString = useExactToString;
}
```

@knutwannheden @timtebeek  what do you think?

Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:59]: There's some serialization that factors in here as well, where we can't be certain the default constructor is called. In other places having a Boolean field then allows us to differentiate, and perhaps using that uniformly would be best then here as well.

Comment (knutwannheden) [reply src/main/java/org/openrewrite/java/migrate/lombok/LombokValueToRecord.java:59]: While it may seem a bit odd, I also think it is more important that we stay consistent with respect to other recipes. In the future we may be able to find a better solution for this. But as Tim says, it has to do with the serialization in the SaaS platform.

**Comment by timtebeek** [comment]
Thanks a lot @kevin0x90 ! You can try out this change through [our snapshot versions](https://docs.openrewrite.org/reference/snapshot-instructions), pending the next release.

I'll deploy it to https://app.moderne.io as well, and see how it performs against our own code base as well.

**Comment by timtebeek** [comment]
Ran it on a first few projects; think we might need some minor tweaks and polish. Example diff below
image

Two quick findings:
1. the fields converted to constructor args lack the corresponding indentation
2. we remove the `@Value` annotation anywhere in the compilation unit, not just on the record

**Comment by timtebeek** [comment]
Follow up now in:
- https://github.com/openrewrite/rewrite-migrate-java/pull/334

**Pull request merged**
